misk/src/main/kotlin/misk/web/jetty/ConnectionListener.kt

```diff
@@ -18,24 +19,28 @@ internal class ConnectionListener(
   private val totalMessagesSent = AtomicLong()
 
   private val connections = Collections.newSetFromMap(ConcurrentHashMap<ConnectionKey, Boolean>())
+  private val openedConnections = ConcurrentLinkedQueue<ConnectionKey>()
+  private val closedConnections = ConcurrentLinkedQueue<ConnectionKey>()
   private val labels = ConnectionMetrics.forPort(protocol, port)
 
+  // Updating prometheus metrics requires monitor locks and its best to avoid acquiring those
+  // with the high number of jetty threads. Instead we queue up changes to the connection and
+  // process the changes when metrics are refreshed by a single background thread.
   override fun onOpened(connection: Connection) {
-    connections.add(ConnectionKey(connection))
-    metrics.activeConnections.labels(*labels).inc()
-    metrics.acceptedConnections.labels(*labels).inc()
+    openedConnections.add(ConnectionKey(connection))
   }
 
   override fun onClosed(connection: Connection) {
-    // Force a refresh so that we gather the final stats on the connection
-    refreshMetrics()
```

Comment:
to clarify I don't think the problem is really Prometheus locking to update metrics. The real problem is we were iterating over every connection (maybe 1000s) every time a connection is closed. That's a lot of contention for the same metrics.